### ISSUE TITLE
 Proposal: simplify typed N-D array storage

### ISSUE THREAD

**Issue by fangq** [msgpack/msgpack#268]
see https://github.com/msgpack/msgpack/pull/267

A simple header to speedup reading/writing large N-D arrays and save space. Particularly useful when storing 2D image data and 3D or high dimensional data form scientific research.

A similar array header was supported by [UBJSON](http://ubjson.org/type-reference/container-types/#optimized-format) and its extensions to N-D array in the [JData specification draft](https://github.com/fangq/jdata/blob/master/JData_specification.md#binary-jdata-storage-grammar). A similar and strong need from the UBJSON user community of such feature was discussed previously [here](https://github.com/ubjson/universal-binary-json/issues/61).

I want to mention that this feature can also result in significant space saving. For example, for storing a 1000x2 0-1 valued array using MessagePack requires ~3,003 bytes, with this format, it requires 1+1+1+2*4+1000*2= 2,011 bytes, that's over 30% space saving.

**Comment by fangq** [comment]
Just to explain this a little bit more, I propose to use the unused label `0xc1` to represent a new array class that allows one to store N-D numerical arrays efficiently. The proposed construct is the `0xc1` byte followed a 1-byte type, 1-byte dimension number (`D`), followed by `D*4` bytes of `uint32` numbers for the dimensions, then followed by serialized binary data from N-D array (in **column-major** order - like in MATLAB/FORTRAN).

For example, a 3x4x100 float32 array shall be stored as

`[0xc0] [0xca] [3] [___3] [___4] [___100] [F(1,1,1), F(2,1,1), ..., F(3,4,100)]`

The detailed proposal can be found below 

```
typed N-D array stores an array with a specified data type and lengths in upto (2^8)-1 dimensions
+--------+--------+--------+====================================+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
|  0xc1  |  type  |  dim   | dim uint32 integers (N1,N2,...,ND) |   N1*N2*...*ND values of type |
+--------+--------+--------+====================================+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
```
where

```
    * type is a big-endian 8-bit unsigned integer, specifying the data type of the array
    * if type is between 0xca and 0xd3, the specified type can be found in the Summary section above
    * other type values are reserved for future extension of this specification
    * dim is a big-endian 8-bit unsigned integer, specifying the dimensions of the array
    * N1, N2, ..., ND are big-endian uint32 integers, specifying the length of the array along each of the D dimensions
    * the binary data of the N-D array is serialized in the column-major order, i.e. the left-most index is the innermost
```

The decision to use `uint32` format for the dimension vector is to keep the format simple, while having the capacity to store large variety of data. If it is desirable to further reduce the header size, it is possible to add a "dim type" byte after the "type" byte, and then one can replace uint32 N1/N2/...Nd numbers to other integer types, such as uint8.

**Comment by fangq** [comment]
I also suggest to use `[0xc1] [0xc3=true]` and `[0xc1][0xc2=false]` pair to define an array with dynamic length, i.e.
```
[ 0xc1 ] [ 0xc3 ] [type1] [data1] [type2] [data2] ... [type N] [data N] [0xc1] [0xc2]
```
MessagePack constructs always require a length in the container (array, map) header - that means when the header is transmitted, the element count must be known as prior. This can be a problem in some scenarios such as data streaming. In comparison, JSON's "`[...]/{...}`" and UBJSON's non-optimized array/object constructs "`[...]/{...}`" do not have to set the element length at the start of the container. Permitting this could be beneficial for some applications.

Similarly, you may use `[0xc1] [0x0]` and `[0xc1][0x1]` pairs to define dynamic maps.

Let me know what you think.

**Comment by fangq** [comment]
Just want to let you know I decided to switch from column-major to row-major order when serializing an N-D packed array using the proposed array construct. This change was a result of the [updates](https://github.com/fangq/jdata/commit/6f2b64429dccf3637eb344b0d7dfb55171b838dc) to [JData specification Draft 1](https://github.com/fangq/jdata/blob/Draft_1/JData_specification.md).

A [patch](https://github.com/msgpack/msgpack/pull/267/commits/295b1abc2d64d16cde10630f9b816e28b1a46388) was submitted and included in the Pull request #267.

**Comment by zougloub** [comment]
Using 0xc1 for a new type removes the ability to use it as an *escape character* for other possibilities (potential preferred extensions). You did realize that with your dynamic length proposal.

The idea of supporting streaming data should be in another bug (and yeah it would require 0xc1's intervention).

For the ND-array, IMHO you're better off with proposing an extension number -- with dims being a msgpack-serialized int, which would make more compact short arrays -- rather than trying to have this as a core type: msgpack can already represent the same data (maybe you should demonstrate the storage win (and consider further compression of msgpack serializations)), and the proposal isn't flexible enough to represent arbitrary element types, some of which we would want to have in an ND-array (eg. structures, strings, bigints).

Alternatively to the simple typecode, since the all elements are required to be of uniform type, perhaps you could serialize a reference template element metadata completely (considering max. element size), then dump only the (potentially padded) payload of the elements, eliminating all the redundant metadata. That's a complex rule but would allow to store more types of elements, and in the case of int/float it would result in the same thing as the simple typecode.

**Comment by fangq** [comment]
> Using 0xc1 for a new type removes the ability to use it as an _escape character_ for other possibilities (potential preferred extensions). You did realize that with your dynamic length proposal.

for the streaming proposal (which I did not include it in this pull request), yes, that is a dynamic length case, but for the ND-array construct, I don't think it is dynamic length - before reading any actual data payload, you always know the length and type. 

to me, it is difficult to use 0xc1 for "escaping" purposes, because it does not have a length, so you don't know when is the end of the escape. One can skip 0xc1 one at a time, but I think, maybe I am wrong, it is not really possible to escape a block of bytes. That makes it less useful.



> The idea of supporting streaming data should be in another bug (and yeah it would require 0xc1's intervention).

happy to do that.


> For the ND-array, IMHO you're better off with proposing an extension number -- with dims being a msgpack-serialized int, which would make more compact short arrays -- rather than trying to have this as a core type: msgpack can already represent the same data (maybe you should demonstrate the storage win (and consider further compression of msgpack serializations)), and the proposal isn't flexible enough to represent arbitrary element types, some of which we would want to have in an ND-array (eg. structures, strings, bigints).


whether ND-array should be considered as a core-type or not, it really depends on applications. For what I do (research data processing), ND-array is essential, sometimes more fundamental than many other complex types. 

the idea of defining a type template to deal with composite record has been discussed in the below UBJSON trackers

https://github.com/ubjson/universal-binary-json/issues/43
https://github.com/ubjson/universal-binary-json/issues/60
https://github.com/ubjson/universal-binary-json/issues/50

some of these have been implemented in ubjson. the support for composite records can be quite complex and deserve much discussions, but I think in the case of ND array of simple types, the concept and utility is quite straightforward, and I don't think it should wait until the template support is fully cooked.


> 
> Alternatively to the simple typecode, since the all elements are required to be of uniform type, perhaps you could serialize a reference template element metadata completely (considering max. element size), then dump only the (potentially padded) payload of the elements, eliminating all the redundant metadata. That's a complex rule but would allow to store more types of elements, and in the case of int/float it would result in the same thing as the simple typecode.

what I want to highlight here is that my proposed format does not conflict with the future extension towards template support. Right now, I only reserved **0xc1 + numerical types (0xca-0xd3)** for this ND packed array proposal. The ability to extend to support more complex records can attach the container markers (arrays, maps) with 0xc1. Think of this proposal as the simplest template.

**Comment by trulede** [comment]
Can you do this with the EXT?

```
+--------+--------+--------+--------+========+
|  0xc8  |YYYYYYYY|YYYYYYYY|  type  |  data  |
+--------+--------+--------+--------+========+
```

0xC8 = EXT 16
YYY... = N bytes
type = Float32_3 (custom type indicating data type and dimension of 3)
Array dimension = N / 4 / 3

But actually, it would be easier to use an array of EXTs, one object for each dimension, as follows:

`[ 10010011 | [ 0xC8 | YY... | 0xCA | data ][ 0xC8 | YY... | 0xCA | data ][ 0xC8 | YY... | 0xCA | data ] ]
`
10010011 = array with 3 objects, one for each dimension.

I may use a similar technique for encoding sparse arrays, or arrays with mixed types where I expect a longer sequence of one type or another, so such an encoding will be more memory efficient. Essentially using EXT to encode the data, and an enclosing array to define additional "instructions" for decoding the data.

**Comment by codeinred** [comment]
Just want to comment - there's continued interest in packed arrays and N-D array types

**Comment by zougloub** [comment]
Yeah. Note that CBOR has a few registered and unregistered extensions for this.
